game-frontend/src/components/user-script/monaco-helper.ts

```diff
@@ -0,0 +1,46 @@
+import { useState, useEffect, useRef } from 'react';
+import { monaco, EditorDidMount } from '@monaco-editor/react';
+import monacoEditor from 'monaco-editor/esm/vs/editor/editor.api';
+
+export type MonacoInstance = typeof monacoEditor;
+type EditorRef = React.MutableRefObject<
+  monacoEditor.editor.IStandaloneCodeEditor | undefined
+>;
+export const useHandleEditorRef = (): [EditorRef, EditorDidMount] => {
+  const editorRef = useRef<monacoEditor.editor.IStandaloneCodeEditor>();
+  const didMount: EditorDidMount = (_, editor) => {
+    editorRef.current = editor;
+  };
+  return [editorRef, didMount];
+};
+
+export const useMonacoInstance = (): MonacoInstance | undefined => {
+  const [monacoInstance, setMonacoInstance] = useState<MonacoInstance>();
+  useEffect(() => {
+    monaco.init().then(m => {
+      setMonacoInstance(m);
+    });
+  }, []);
+  return monacoInstance;
+};
+
+export const useSaveHotKey = (editorRef: EditorRef, save: () => void): void => {
+  const monacoInstance = useMonacoInstance();
+
+  useEffect(() => {
+    if (editorRef && editorRef.current && monacoInstance) {
+      const m = monacoInstance as MonacoInstance;
+      editorRef.current.addAction({
```

Comment:
In the current implementation every code change in the editor will retrigger `addAction`.
It does work, but it's certainly not the proper way to handle this kind of behavior.
Consider refactoring it to reduce the number of unnecessary calls on a foreign API?

Reply:
This is resolved here: https://github.com/code-and-conquer/darwin/pull/56